README.md
<p align=center>
  <a href="https://github.com/TimoKunze/ToolBarControls/releases/tag/1.3.5">
    <img alt="Release 1.3.5 Release" src="https://img.shields.io/badge/release-1.3.5-0688CB.svg">
  </a>
  <a href="https://github.com/TimoKunze/ToolBarControls/releases">
    <img alt="Download ToolBarControls" src="https://img.shields.io/badge/download-latest-0688CB.svg">
  </a>
  <a href="https://github.com/TimoKunze/ToolBarControls/blob/master/LICENSE">
    <img alt="License" src="https://img.shields.io/badge/license-MIT-0688CB.svg">
  </a>
  <a href="https://www.paypal.com/xclick/business=TKunze71216%40gmx.de&item_name=ToolBarControls&no_shipping=1&tax=0&currency_code=EUR">
    <img alt="Donate" src="https://img.shields.io/badge/%24-donate-E44E4A.svg">
  </a>
</p>

# ToolBarControls
An ActiveX control library for Visual Basic 6 that contains a rebar and a tool bar control.

I've developed this ActiveX control library between 2010 and 2013 and did update it on a regular basis until 2016. Currently I have little interest to maintain this project any longer, but I think the code might be of some use to others.

# Before you make changes
If you make changes to the code and deploy the binary, keep in mind that ActiveX controls are COM components and therefore should stay binary compatible as long as you don't change the COM object's, i.e. the ActiveX control's public class name and GUIDs. Otherwise people using these components are likely to end up in the famous COM hell.

# Dependencies
You will need the [Microsoft Windows 10 SDK](https://developer.microsoft.com/en-us/windows/downloads/windows-10-sdk), ATL, and [WTL 10](https://sourceforge.net/projects/wtl/).

## Required Change to ATL
Some version of ATL have a bug in ```AtlIPersistPropertyBag_Load``` which causes crashes. In the file atlcom.h search for ```AtlIPersistPropertyBag_Load```. Inside the implementation of this function search for ```HRESULT hr = pPropBag->Read(pMap[i].szDesc, &var, pErrorLog);```. Make sure that there is this code before this line: ```var.pdispVal = NULL;```.


CHANGELOG
Version 1.3.5 - 28/04/2019

- FIX: Dragging data to certain programs like Total Commander or FileZilla did not work if the data was provided only in the OLESetData event.



Version 1.3.4 - 04/02/2018

- Upgraded to WTL 10.
- Integrated the GitHub repository into the documentation and the About dialog.



Version 1.3.3 - 16/01/2018

- Project published on GitHub.



Version 1.3.2 - 31/07/2016

- Fixed some memory leaks that occurred when firing certain events.



Version 1.3.1 - 26/01/2016

- FIX: The controls could crash during OLE drag'n'drop.
- FIX: The controls could crash the other process during OLE drag'n'drop, especially when IOLEDataObject::SetDropDescription() had been called.



Version 1.3.0 - 24/01/2016

- Added the FinishOLEDragDrop method to each control. It should not be used if not really needed. It can be used to hide the drag image if execution of the OLEDragDrop event takes a long time or if a message is displayed from within this event.
- Initial support for Windows 10.



Version 1.2.4 - 24/06/2015

- Fixed a problem with the CustomDraw event that could crash the control.
- Fixed a problem with the Align property that occurred if the control's x or y coordinate has been 0.



Version 1.2.3 - 02/11/2014

- Fixed the controls' Alignment property. Now it behaves like the Alignment property of VB's intrinsic controls.



Version 1.2.2 - 18/10/2014

- Fixed a crash that could occur in VB6 IDE when changing the ReBar control's Appearance property.
- Added a new sample "PNG Icons" which demonstrates how to load tool bar icons from a PNG file using GDI+.



Version 1.2.1 - 10/08/2014

- FIX: The tool bar control's *Click events have not been raised for the chevron popup window.
- FIX: The IToolBar::MousePointer property did not work properly for the chevron popup window.



Version 1.2.0 - 21/07/2014

- Added the XClick, and XDblClick events for support of the extended mouse buttons.
- Extended the Mouse* events to support the extended mouse buttons.
- FIX: The Font and MouseIcon properties could leak GDI resources.
- Initial support for Windows 8.1.



Version 1.1.2 - 09/05/2013

- Fixed compatibility of the ToolBar control with pager controls.
- FIX: IToolBar::BackStyle = bksTransparent did work only partially.



Version 1.1.1 - 13/01/2013

- FIX: The last character of the band or button text could not be used as accelerator key.
- FIX: The ResizedControlWindow event was not fired when it should have been fired.



Version 1.1.0 - 26/08/2012

- Fixed IControlHostWindow::CalculateWindowSize() and IControlHostWindow::Move(). Both methods did not handle default values correctly.
- Initial support for Windows 8.



Version 1.0.3 - 25/03/2012

- FIX: The background of transparent stand-alone tool bars that used the classic theme, wasn't drawn correctly. This wasn't really a bug because everything was working as designed. However, in most cases this wasn't the expected result. I "fixed" this problem by detecting the situation, forcing the RaiseCustomDrawEventOnEraseBackground property to True and drawing the background after the CustomDraw event (for drawStage = cdsPreErase) if the return value doesn't include the cdrvSkipDefault flag.
- Added a new sample "Breadcrumb Bar" which demonstrates how to use the both controls together with ComboListBoxControls and the ProgressBar control to build a neat breadcrumb bar like that of Windows Explorer.
- Reimplemented the subclassing code of the VB6 samples to make it more robust.



Version 1.0.2 - 08/01/2012

- FIX: On systems older than Windows 7, the tool bar chevron popup window could be displayed off-screen.
- FIX: The GetAvailableButton event had problems with the button texts. This issue could not be fixed entirely, but now a simple workaround exists. See the help topic about this event for details.
- FIX: The IToolBar::LoadToolBarStateFromRegistry method did not work properly.
- Fixed a memory leak that occurred during drag'n'drop if Aero drag-images have been used.



Version 1.0.1 - 28/11/2011


src/DispIDs.h
// DispIDs.h: Defines a DISPID for each COM property and method we're providing.

// IReBar
// properties
#define DISPID_RB_ALLOWBANDREORDERING													1
#define DISPID_RB_APPEARANCE																	2
#define DISPID_RB_AUTOUPDATELAYOUT														3
#define DISPID_RB_BACKCOLOR																		4
#define DISPID_RB_BANDS																				5
#define DISPID_RB_BORDERSTYLE																	6
#define DISPID_RB_CONTROLHEIGHT																7
#define DISPID_RB_DISABLEDEVENTS															8
#define DISPID_RB_DISPLAYBANDSEPARATORS												9
#define DISPID_RB_DISPLAYSPLITTER															10
#define DISPID_RB_DONTREDRAW																	11
#define DISPID_RB_ENABLED																			DISPID_ENABLED
#define DISPID_RB_FIXEDBANDHEIGHT															12
#define DISPID_RB_FONT																				13
#define DISPID_RB_FORECOLOR																		14
#define DISPID_RB_HIGHLIGHTCOLOR															15
#define DISPID_RB_HIMAGELIST																	16
#define DISPID_RB_HPALETTE																		17
#define DISPID_RB_HOVERTIME																		18
#define DISPID_RB_HWND																				DISPID_VALUE
#define DISPID_RB_HWNDNOTIFICATIONRECEIVER										19
//#define DISPID_RB_HWNDTOOLTIP																	20
#define DISPID_RB_MDIFRAMEMENUBAND														21
#define DISPID_RB_MOUSEICON																		22
#define DISPID_RB_MOUSEPOINTER																23
#define DISPID_RB_NATIVEDROPTARGET														24
#define DISPID_RB_NUMBEROFROWS																25
#define DISPID_RB_ORIENTATION																	26
#define DISPID_RB_REGISTERFOROLEDRAGDROP											27
#define DISPID_RB_REPLACEMDIFRAMEMENU													28
#define DISPID_RB_RIGHTTOLEFT																	29
#define DISPID_RB_SHADOWCOLOR																	30
#define DISPID_RB_SUPPORTOLEDRAGIMAGES												31
#define DISPID_RB_TOGGLEONDOUBLECLICK													32
#define DISPID_RB_USESYSTEMFONT																33
#define DISPID_RB_VERSION																			34
#define DISPID_RB_VERTICALSIZINGGRIPSONVERTICALORIENTATION		35
// fingerprint
#define DISPID_RB_APPID																				500
#define DISPID_RB_APPNAME																			501
#define DISPID_RB_APPSHORTNAME																502
#define DISPID_RB_BUILD																				503
#define DISPID_RB_CHARSET																			504
#define DISPID_RB_ISRELEASE																		505
#define DISPID_RB_PROGRAMMER																	506
#define DISPID_RB_TESTER																			507
// methods
#define DISPID_RB_ABOUT																				DISPID_ABOUTBOX
#define DISPID_RB_BEGINDRAGBAND																36
#define DISPID_RB_DRAGMOVEBAND																37
#define DISPID_RB_ENDDRAGBAND																	38
#define DISPID_RB_GETMARGINS																	39
#define DISPID_RB_HITTEST																			40
#define DISPID_RB_LOADSETTINGSFROMFILE												41
#define DISPID_RB_REFRESH																			DISPID_REFRESH
#define DISPID_RB_SAVESETTINGSTOFILE													42
#define DISPID_RB_SIZETORECTANGLE															43
#define DISPID_RB_FINISHOLEDRAGDROP														44


// _IReBarEvents
// methods
#define DISPID_RBE_AUTOBREAKINGBAND														1
#define DISPID_RBE_AUTOSIZED																	2
#define DISPID_RBE_BANDBEGINDRAG															3
#define DISPID_RBE_BANDBEGINRDRAG															4
#define DISPID_RBE_BANDENDDRAG																5
#define DISPID_RBE_BANDMOUSEENTER															6
#define DISPID_RBE_BANDMOUSELEAVE															7
#define DISPID_RBE_BEFOREDISPLAYMDICHILDSYSTEMMENU						8
#define DISPID_RBE_CHEVRONCLICK																9
#define DISPID_RBE_CLEANUPMDICHILDSYSTEMMENU									10
#define DISPID_RBE_CLICK																			11
#define DISPID_RBE_CONTEXTMENU																12
#define DISPID_RBE_CUSTOMDRAW																	13
#define DISPID_RBE_DBLCLICK																		14
#define DISPID_RBE_DESTROYEDCONTROLWINDOW											15
#define DISPID_RBE_DRAGGINGSPLITTER														16
#define DISPID_RBE_FREEBANDDATA																17
#define DISPID_RBE_HEIGHTCHANGED															18
#define DISPID_RBE_INSERTEDBAND																19
#define DISPID_RBE_INSERTINGBAND															20
#define DISPID_RBE_LAYOUTCHANGED															21
#define DISPID_RBE_MCLICK																			22
#define DISPID_RBE_MDBLCLICK																	23
#define DISPID_RBE_MOUSEDOWN																	24
#define DISPID_RBE_MOUSEENTER																	25
#define DISPID_RBE_MOUSEHOVER																	26
#define DISPID_RBE_MOUSELEAVE																	27
#define DISPID_RBE_MOUSEMOVE																	28
#define DISPID_RBE_MOUSEUP																		29
#define DISPID_RBE_NONCLIENTHITTEST														30
#define DISPID_RBE_OLEDRAGDROP																31
#define DISPID_RBE_OLEDRAGENTER																32
#define DISPID_RBE_OLEDRAGLEAVE																33
#define DISPID_RBE_OLEDRAGMOUSEMOVE														34
#define DISPID_RBE_RAWMENUMESSAGE															35
#define DISPID_RBE_RCLICK																			36
#define DISPID_RBE_RDBLCLICK																	37
#define DISPID_RBE_RECREATEDCONTROLWINDOW											38
#define DISPID_RBE_RELEASEDMOUSECAPTURE												39
#define DISPID_RBE_REMOVEDBAND																40
#define DISPID_RBE_REMOVINGBAND																41
#define DISPID_RBE_RESIZEDCONTROLWINDOW												42
#define DISPID_RBE_RESIZINGCONTAINEDWINDOW										43
#define DISPID_RBE_SELECTEDMENUITEM														44
#define DISPID_RBE_TOGGLINGBAND																45
#define DISPID_RBE_XCLICK																			46
#define DISPID_RBE_XDBLCLICK																	47


// IReBarBand
// properties
#define DISPID_RBB_ADDMARGINSAROUNDCHILD											1
#define DISPID_RBB_BACKCOLOR																	2
#define DISPID_RBB_BANDDATA																		3
#define DISPID_RBB_CHEVRONBUTTONOBJECTSTATE										4
#define DISPID_RBB_CURRENTHEIGHT															5
#define DISPID_RBB_CURRENTWIDTH																6
#define DISPID_RBB_FIXEDBACKGROUNDBITMAPORIGIN								7
#define DISPID_RBB_FORECOLOR																	8
#define DISPID_RBB_HBACKGROUNDBITMAP													9
#define DISPID_RBB_HCONTAINEDWINDOW														DISPID_VALUE
#define DISPID_RBB_HEIGHTCHANGESTEPSIZE												10
#define DISPID_RBB_HIDEIFVERTICAL															11
#define DISPID_RBB_ICONINDEX																	12
#define DISPID_RBB_ID																					13
#define DISPID_RBB_IDEALWIDTH																	14
#define DISPID_RBB_INDEX																			15
#define DISPID_RBB_KEEPINFIRSTROW															16
#define DISPID_RBB_MAXIMUMHEIGHT															17
#define DISPID_RBB_MINIMUMHEIGHT															18
#define DISPID_RBB_MINIMUMWIDTH																19
#define DISPID_RBB_NEWLINE																		20
#define DISPID_RBB_RESIZABLE																	21
#define DISPID_RBB_ROWHEIGHT																	22
#define DISPID_RBB_SHOWTITLE																	23
#define DISPID_RBB_SIZINGGRIPVISIBILITY												24
#define DISPID_RBB_TEXT																				25
#define DISPID_RBB_TITLEWIDTH																	26
#define DISPID_RBB_USECHEVRON																	27
#define DISPID_RBB_VARIABLEHEIGHT															28
#define DISPID_RBB_VISIBLE																		29
#define DISPID_RBB_CHEVRONHOT																	38
#define DISPID_RBB_CHEVRONPUSHED															39
#define DISPID_RBB_CHEVRONVISIBLE															40
// methods
#define DISPID_RBB_CLICKCHEVRON																30
#define DISPID_RBB_GETBORDERSIZES															31
#define DISPID_RBB_GETCHEVRONRECTANGLE												32
#define DISPID_RBB_GETRECTANGLE																33
#define DISPID_RBB_HIDE																				34
#define DISPID_RBB_MAXIMIZE																		35
#define DISPID_RBB_MINIMIZE																		36
#define DISPID_RBB_SHOW																				37


// IReBarBands
// properties
#define DISPID_RBBS_CASESENSITIVEFILTERS											1
#define DISPID_RBBS_COMPARISONFUNCTION												2
#define DISPID_RBBS_FILTER																		3
#define DISPID_RBBS_FILTERTYPE																4
#define DISPID_RBBS_ITEM																			DISPID_VALUE
#define DISPID_RBBS__NEWENUM																	DISPID_NEWENUM
// methods
#define DISPID_RBBS_ADD																				5
#define DISPID_RBBS_CONTAINS																	6
#define DISPID_RBBS_COUNT																			7
#define DISPID_RBBS_REMOVE																		8
#define DISPID_RBBS_REMOVEALL																	9


// IVirtualReBarBand
// properties
#define DISPID_VRBB_ADDMARGINSAROUNDCHILD											1
#define DISPID_VRBB_BACKCOLOR																	2
#define DISPID_VRBB_BANDDATA																	3
#define DISPID_VRBB_CHEVRONBUTTONOBJECTSTATE									4
#define DISPID_VRBB_CURRENTHEIGHT															5
#define DISPID_VRBB_CURRENTWIDTH															6
#define DISPID_VRBB_FIXEDBACKGROUNDBITMAPORIGIN								7
#define DISPID_VRBB_FORECOLOR																	8
#define DISPID_VRBB_HBACKGROUNDBITMAP													9
#define DISPID_VRBB_HCONTAINEDWINDOW													DISPID_VALUE
#define DISPID_VRBB_HEIGHTCHANGESTEPSIZE											10
#define DISPID_VRBB_HIDEIFVERTICAL														11
#define DISPID_VRBB_ICONINDEX																	12
#define DISPID_VRBB_ID																				13
#define DISPID_VRBB_IDEALWIDTH																14
#define DISPID_VRBB_INDEX																			15
#define DISPID_VRBB_KEEPINFIRSTROW														16
#define DISPID_VRBB_MAXIMUMHEIGHT															17
#define DISPID_VRBB_MINIMUMHEIGHT															18
#define DISPID_VRBB_MINIMUMWIDTH															19
#define DISPID_VRBB_NEWLINE																		20
#define DISPID_VRBB_RESIZABLE																	21
#define DISPID_VRBB_SHOWTITLE																	22
#define DISPID_VRBB_SIZINGGRIPVISIBILITY											23
#define DISPID_VRBB_TEXT																			24
#define DISPID_VRBB_TITLEWIDTH																25
#define DISPID_VRBB_USECHEVRON																26
#define DISPID_VRBB_VARIABLEHEIGHT														27
#define DISPID_VRBB_VISIBLE																		28
// methods
#define DISPID_VRBB_GETCHEVRONRECTANGLE												29


// IToolBar
// properties
#define DISPID_TB_ALLOWCUSTOMIZATION													1
#define DISPID_TB_ALWAYSDISPLAYBUTTONTEXT											2
#define DISPID_TB_ANCHORHIGHLIGHTING													3
#define DISPID_TB_APPEARANCE																	4
#define DISPID_TB_BACKSTYLE																		5
#define DISPID_TB_BORDERSTYLE																	6
#define DISPID_TB_BUTTONHEIGHT																7
#define DISPID_TB_BUTTONS																			8
#define DISPID_TB_BUTTONROWCOUNT															9
#define DISPID_TB_BUTTONSTYLE																	10
#define DISPID_TB_BUTTONTEXTPOSITION													11
#define DISPID_TB_BUTTONWIDTH																	12
#define DISPID_TB_COMMANDENABLED															13
#define DISPID_TB_DISABLEDEVENTS															14
#define DISPID_TB_DISPLAYMENUDIVIDER													15
#define DISPID_TB_DISPLAYPARTIALLYCLIPPEDBUTTONS							16
#define DISPID_TB_DONTREDRAW																	17
#define DISPID_TB_DRAGCLICKTIME																18
#define DISPID_TB_DRAGDROPCUSTOMIZATIONMODIFIERKEY						19
#define DISPID_TB_DRAGGEDBUTTONS															20
#define DISPID_TB_DROPDOWNGAP																	21
#define DISPID_TB_ENABLED																			DISPID_ENABLED
#define DISPID_TB_FIRSTBUTTONINDENTATION											22
#define DISPID_TB_FOCUSONCLICK																23
#define DISPID_TB_FONT																				24
#define DISPID_TB_HCHEVRONMENU																25
//#define DISPID_TB_HDRAGIMAGELIST															26
#define DISPID_TB_HIGHLIGHTCOLOR															27
#define DISPID_TB_HIMAGELIST																	28
#define DISPID_TB_HORIZONTALBUTTONPADDING											29
#define DISPID_TB_HORIZONTALBUTTONSPACING											30
#define DISPID_TB_HORIZONTALICONCAPTIONGAP										31
#define DISPID_TB_HORIZONTALTEXTALIGNMENT											32
#define DISPID_TB_HOTBUTTON																		33
#define DISPID_TB_HOVERTIME																		34
#define DISPID_TB_HWND																				35
#define DISPID_TB_HWNDCHEVRONTOOLBAR													36
#define DISPID_TB_HWNDTOOLTIP																	37
#define DISPID_TB_IDEALHEIGHT																	38
#define DISPID_TB_IDEALWIDTH																	39
#define DISPID_TB_IMAGELISTCOUNT															40
#define DISPID_TB_INSERTMARKCOLOR															41
#define DISPID_TB_MAXIMUMBUTTONWIDTH													42
#define DISPID_TB_MAXIMUMHEIGHT																43
#define DISPID_TB_MAXIMUMTEXTROWS															44
#define DISPID_TB_MAXIMUMWIDTH																45
#define DISPID_TB_MENUBARTHEME																46
#define DISPID_TB_MENUMODE																		47
#define DISPID_TB_MINIMUMBUTTONWIDTH													48
#define DISPID_TB_MOUSEICON																		49
#define DISPID_TB_MOUSEPOINTER																50
#define DISPID_TB_MULTICOLUMN																	51
#define DISPID_TB_NATIVEDROPTARGET														52
#define DISPID_TB_NORMALDROPDOWNBUTTONSTYLE										53
#define DISPID_TB_OLEDRAGIMAGESTYLE														54
#define DISPID_TB_ORIENTATION																	55
#define DISPID_TB_PROCESSCONTEXTMENUKEYS											56
#define DISPID_TB_RAISECUSTOMDRAWEVENTONERASEBACKGROUND				57
#define DISPID_TB_REGISTERFOROLEDRAGDROP											58
#define DISPID_TB_RIGHTTOLEFT																	59
#define DISPID_TB_SHADOWCOLOR																	60
#define DISPID_TB_SHOWDRAGIMAGE																61
#define DISPID_TB_SHOWTOOLTIPS																62
#define DISPID_TB_SUGGESTEDICONSIZE														63
#define DISPID_TB_SUPPORTOLEDRAGIMAGES												64
#define DISPID_TB_USEMNEMONICS																65
#define DISPID_TB_USESYSTEMFONT																66
#define DISPID_TB_VERSION																			67
#define DISPID_TB_VERTICALBUTTONPADDING												68
#define DISPID_TB_VERTICALBUTTONSPACING												69
#define DISPID_TB_VERTICALTEXTALIGNMENT												70
#define DISPID_TB_WRAPBUTTONS																	71
// fingerprint
#define DISPID_TB_APPID																				500
#define DISPID_TB_APPNAME																			501
#define DISPID_TB_APPSHORTNAME																502
#define DISPID_TB_BUILD																				503
#define DISPID_TB_CHARSET																			504
#define DISPID_TB_ISRELEASE																		505
#define DISPID_TB_PROGRAMMER																	506
#define DISPID_TB_TESTER																			507
// methods
#define DISPID_TB_ABOUT																				DISPID_ABOUTBOX
#define DISPID_TB_AUTOSIZE																		72
#define DISPID_TB_COUNTACCELERATOROCCURRENCES									73
#define DISPID_TB_CREATEBUTTONCONTAINER												74
#define DISPID_TB_CUSTOMIZE																		75
#define DISPID_TB_DISPLAYCHEVRONPOPUPWINDOW										76
#define DISPID_TB_GETCLOSESTINSERTMARKPOSITION								77
#define DISPID_TB_GETCOMMANDFORHOTKEY													78
#define DISPID_TB_GETINSERTMARKPOSITION												79
#define DISPID_TB_HITTEST																			80
#define DISPID_TB_HITTESTCHEVRONTOOLBAR												81
#define DISPID_TB_LOADDEFAULTIMAGES														82
#define DISPID_TB_LOADSETTINGSFROMFILE												83
#define DISPID_TB_LOADTOOLBARSTATEFROMREGISTRY								84
#define DISPID_TB_OLEDRAG																			85
#define DISPID_TB_REFRESH																			DISPID_REFRESH
#define DISPID_TB_REGISTERHOTKEY															86
#define DISPID_TB_SAVESETTINGSTOFILE													87
#define DISPID_TB_SAVETOOLBARSTATETOREGISTRY									88
#define DISPID_TB_SETBUTTONROWCOUNT														89
#define DISPID_TB_SETHOTBUTTON																90
#define DISPID_TB_SETINSERTMARKPOSITION												91
#define DISPID_TB_UNREGISTERHOTKEY														92
#define DISPID_TB_FINISHOLEDRAGDROP														93


// _IToolBarEvents
// methods
#define DISPID_TBE_BEFOREDISPLAYCHEVRONPOPUP									1
#define DISPID_TBE_BEGINCUSTOMIZATION													2
#define DISPID_TBE_BUTTONBEGINDRAG														3
#define DISPID_TBE_BUTTONBEGINRDRAG														4
#define DISPID_TBE_BUTTONGETDISPLAYINFO												5
#define DISPID_TBE_BUTTONGETDROPDOWNMENU											6
#define DISPID_TBE_BUTTONGETINFOTIPTEXT												7
#define DISPID_TBE_BUTTONMOUSEENTER														8
#define DISPID_TBE_BUTTONMOUSELEAVE														9
#define DISPID_TBE_BUTTONSELECTIONSTATECHANGED								10
#define DISPID_TBE_CLICK																			11
#define DISPID_TBE_CONTEXTMENU																12
#define DISPID_TBE_CUSTOMDRAW																	13
#define DISPID_TBE_CUSTOMIZEDCONTROL													14
#define DISPID_TBE_CUSTOMIZEDIALOGREMOVINGBUTTON							15
#define DISPID_TBE_DBLCLICK																		16
#define DISPID_TBE_DESTROYEDCONTROLWINDOW											17
#define DISPID_TBE_DESTROYINGCHEVRONPOPUP											18
#define DISPID_TBE_DISPLAYCUSTOMIZATIONHELP										19
#define DISPID_TBE_DROPDOWN																		20
#define DISPID_TBE_ENDCUSTOMIZATION														21
#define DISPID_TBE_EXECUTECOMMAND															22
#define DISPID_TBE_FREEBUTTONDATA															23
#define DISPID_TBE_GETAVAILABLEBUTTON													24
#define DISPID_TBE_HOTBUTTONCHANGEWRAPPING										25
#define DISPID_TBE_HOTBUTTONCHANGING													26
#define DISPID_TBE_INITIALIZECUSTOMIZATIONDIALOG							27
#define DISPID_TBE_INITIALIZETOOLBARSTATEREGISTRYRESTORAGE		28
#define DISPID_TBE_INITIALIZETOOLBARSTATEREGISTRYSTORAGE			29
#define DISPID_TBE_INSERTEDBUTTON															30
#define DISPID_TBE_INSERTINGBUTTON														31
#define DISPID_TBE_ISDUPLICATEACCELERATOR											32
#define DISPID_TBE_KEYDOWN																		33
#define DISPID_TBE_KEYPRESS																		34
#define DISPID_TBE_KEYUP																			35
#define DISPID_TBE_MAPACCELERATOR															36
#define DISPID_TBE_MCLICK																			37
#define DISPID_TBE_MDBLCLICK																	38
#define DISPID_TBE_MOUSEDOWN																	39
#define DISPID_TBE_MOUSEENTER																	40
#define DISPID_TBE_MOUSEHOVER																	41
#define DISPID_TBE_MOUSELEAVE																	42
#define DISPID_TBE_MOUSEMOVE																	43
#define DISPID_TBE_MOUSEUP																		44
#define DISPID_TBE_OLECOMPLETEDRAG														45
#define DISPID_TBE_OLEDRAGDROP																46
#define DISPID_TBE_OLEDRAGENTER																47
#define DISPID_TBE_OLEDRAGENTERPOTENTIALTARGET								48
#define DISPID_TBE_OLEDRAGLEAVE																49
#define DISPID_TBE_OLEDRAGLEAVEPOTENTIALTARGET								50
#define DISPID_TBE_OLEDRAGMOUSEMOVE														51
#define DISPID_TBE_OLEGIVEFEEDBACK														52
#define DISPID_TBE_OLEQUERYCONTINUEDRAG												53
#define DISPID_TBE_OLERECEIVEDNEWDATA													54
#define DISPID_TBE_OLESETDATA																	55
#define DISPID_TBE_OLESTARTDRAG																56
#define DISPID_TBE_OPENPOPUPMENU															57
#define DISPID_TBE_QUERYINSERTBUTTON													58
#define DISPID_TBE_QUERYREMOVEBUTTON													59
#define DISPID_TBE_RAWMENUMESSAGE															60
#define DISPID_TBE_RCLICK																			61
#define DISPID_TBE_RDBLCLICK																	62
#define DISPID_TBE_RECREATEDCONTROLWINDOW											63
#define DISPID_TBE_REMOVEDBUTTON															64
#define DISPID_TBE_REMOVINGBUTTON															65
#define DISPID_TBE_RESETCUSTOMIZATIONS												66
#define DISPID_TBE_RESIZEDCONTROLWINDOW												67
#define DISPID_TBE_RESTOREBUTTONFROMREGISTRYSTREAM						68
#define DISPID_TBE_SAVEBUTTONTOREGISTRYSTREAM									69
#define DISPID_TBE_SELECTEDMENUITEM														70
#define DISPID_TBE_XCLICK																			71
#define DISPID_TBE_XDBLCLICK																	72


// ToolBarButton
// properties
#define DISPID_TBB_AUTOSIZE																		1
#define DISPID_TBB_BUTTONDATA																	2
#define DISPID_TBB_BUTTONTYPE																	3
#define DISPID_TBB_DISPLAYTEXT																4
#define DISPID_TBB_DROPDOWNSTYLE															5
#define DISPID_TBB_DROPPEDDOWN																6
#define DISPID_TBB_ENABLED																		7
#define DISPID_TBB_FOLLOWEDBYLINEBREAK												8
#define DISPID_TBB_HOT																				9
#define DISPID_TBB_ICONINDEX																	10
#define DISPID_TBB_ID																					DISPID_VALUE
#define DISPID_TBB_IMAGELISTINDEX															11
#define DISPID_TBB_INDEX																			12
#define DISPID_TBB_MARKED																			13
#define DISPID_TBB_PARTOFCHEVRONTOOLBAR												14
#define DISPID_TBB_PARTOFGROUP																15
#define DISPID_TBB_PUSHED																			16
#define DISPID_TBB_SELECTIONSTATE															17
#define DISPID_TBB_SHOULDBEDISPLAYEDINCHEVRONPOPUP						18
#define DISPID_TBB_SHOWINGELLIPSIS														19
#define DISPID_TBB_TEXT																				20
#define DISPID_TBB_USEMNEMONIC																21
#define DISPID_TBB_VISIBLE																		22
#define DISPID_TBB_WIDTH																			23
// methods
#define DISPID_TBB_GETCONTAINEDWINDOW													24
#define DISPID_TBB_GETRECTANGLE																25
#define DISPID_TBB_SETCONTAINEDWINDOW													26


// IToolBarButtons
// properties
#define DISPID_TBBS_CASESENSITIVEFILTERS											1
#define DISPID_TBBS_COMPARISONFUNCTION												2
#define DISPID_TBBS_FILTER																		3
#define DISPID_TBBS_FILTERTYPE																4
#define DISPID_TBBS_ITEM																			DISPID_VALUE
#define DISPID_TBBS__NEWENUM																	DISPID_NEWENUM
// methods
#define DISPID_TBBS_ADD																				5
#define DISPID_TBBS_CONTAINS																	6
#define DISPID_TBBS_COUNT																			7
#define DISPID_TBBS_REMOVE																		8
#define DISPID_TBBS_REMOVEALL																	9


// IToolBarButtonContainer
// properties
#define DISPID_TBBC_ITEM																			DISPID_VALUE
#define DISPID_TBBC__NEWENUM																	DISPID_NEWENUM
// methods
#define DISPID_TBBC_ADD																				1
#define DISPID_TBBC_CLONE																			2
#define DISPID_TBBC_COUNT																			3
#define DISPID_TBBC_CREATEDRAGIMAGE														4
#define DISPID_TBBC_REMOVE																		5
#define DISPID_TBBC_REMOVEALL																	6


// IVirtualToolBarButton
// properties
#define DISPID_VTBB_AUTOSIZE																	1
#define DISPID_VTBB_BUTTONDATA																2
#define DISPID_VTBB_BUTTONTYPE																3
#define DISPID_VTBB_DISPLAYTEXT																4
#define DISPID_VTBB_DROPDOWNSTYLE															5
#define DISPID_VTBB_ENABLED																		6
#define DISPID_VTBB_FOLLOWEDBYLINEBREAK												7
#define DISPID_VTBB_ICONINDEX																	8
#define DISPID_VTBB_ID																				DISPID_VALUE
#define DISPID_VTBB_IMAGELISTINDEX														9
#define DISPID_VTBB_INDEX																			10
#define DISPID_VTBB_MARKED																		11
#define DISPID_VTBB_PARTOFGROUP																12
#define DISPID_VTBB_PUSHED																		13
#define DISPID_VTBB_SELECTIONSTATE														14
#define DISPID_VTBB_SHOWINGELLIPSIS														15
#define DISPID_VTBB_TEXT																			16
#define DISPID_VTBB_USEMNEMONIC																17
#define DISPID_VTBB_VISIBLE																		18
#define DISPID_VTBB_WIDTH																			19
// methods


// IOLEDataObject
// properties
// methods
#define DISPID_ODO_CLEAR																			1
#define DISPID_ODO_GETCANONICALFORMAT													2
#define DISPID_ODO_GETDATA																		3
#define DISPID_ODO_GETFORMAT																	4
#define DISPID_ODO_SETDATA																		5
#define DISPID_ODO_GETDROPDESCRIPTION													6
#define DISPID_ODO_SETDROPDESCRIPTION													7


// IControlHostWindow
// properties
#define DISPID_CHW_HWND																				DISPID_VALUE
#define DISPID_CHW_HWNDCHILD																	1
#define DISPID_CHW_RESIZEABLE																	2
#define DISPID_CHW_VISIBLE																		3
#define DISPID_CHW_WINDOWTITLE																4
// methods
#define DISPID_CHW_CALCULATEWINDOWSIZE												5
#define DISPID_CHW_CREATE																			6
#define DISPID_CHW_DESTROY																		7
#define DISPID_CHW_GETCLIENTRECTANGLE													8
#define DISPID_CHW_GETRECTANGLE																9
#define DISPID_CHW_MOVE																				10


// _IControlHostWindowEvents
// methods
#define DISPID_CHWE_ACTIVATE																	1
#define DISPID_CHWE_CLOSING																		2
#define DISPID_CHWE_CREATED																		3
#define DISPID_CHWE_DEACTIVATE																4
#define DISPID_CHWE_DESTROYED																	5
#define DISPID_CHWE_MOVED																			6
#define DISPID_CHWE_MOVING																		7
#define DISPID_CHWE_RESIZED																		8
#define DISPID_CHWE_RESIZING																	9
#define DISPID_CHWE_TITLEDBLCLICK															10